### PR TITLE
Add minimal LLVM module support for debug information

### LINKED ISSUE #381: Add minimal LLVM module support for debug information
In order to get backtraces with line numbers #6, we need to generate some minimal debug information #7.

This will need some functionality from LLVMDIBuilder.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Add minimal LLVM module support for debug information
Resolves #381

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/print.llvm/llvm.bal [hunk 1]
```diff
@@ -577,6 +605,241 @@ public class FunctionDefn {
...
+                   props.runtimeVersion.toString(), ",", "emissionKind", ":", emissionKindToString.get(props.kind), ")");
+        metadata.addLine(...words);
+
+        metadata.addPreamble("!", "llvm", ".", "dbg", ".", "cu", "=", "!", "{", metadata.ref(), "}");
+        self.compileUnit = metadata;
+        return metadata;
+    }
+
+    // Corresponds to LLVMDIBuilderCreateFile
+    public function createFile(string fileName, string directory) returns Metadata {
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: filename is one word so use `filename` not `fileName`.

File: compiler/modules/print.llvm/llvm.bal [hunk 2]
```diff
@@ -200,6 +201,23 @@ public class Module {
...
+    // Corresponds to LLVMCreateDIBuilder
+    public function createDIBuilder() returns DIBuilder {
+        DIBuilder dIBuilder = new(self);
+        self.dIBuilder = dIBuilder;
+        return dIBuilder;
+    }
+
+    // Corresponds to LLVMAddModuleFlag
+    public function addModuleFlag(ModuleFlagBehavior behavior, ModuleFlag flag) {
+        DIBuilder? dIBuilder = self.dIBuilder;
```
Comment (jclark) [inline compiler/modules/print.llvm/llvm.bal]: This doesn't feel like a good design.

File: compiler/modules/print.llvm/common.bal [hunk 3]
```diff
@@ -103,3 +103,36 @@ public type GlobalProperties record {|
...
+// Corresponds to LLVMDIFlags
+public type DIFlag "zero";
+public type FunctionMetadataProperties record {|
+    Metadata? scope;
+    string? name;
+    string? linkageName;
+    Metadata? file;
+    int lineNo;
+    Metadata? ty;
+    "localToUnit"? localToUnit = "localToUnit";
```
Comment (jclark) [inline compiler/modules/print.llvm/common.bal]: Just use booleans for field values (not the same situation as parameters).

File: compiler/modules/print.llvm/llvm.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: Fixed

File: compiler/modules/print.llvm/llvm.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: Changed such that module will support multiple `DIBuilders` and will create a new one silently if none available.

File: compiler/modules/print.llvm/common.bal (hunk 3 above)
Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/common.bal]: Fixed

File: compiler/modules/print.llvm/llvm.bal (hunk 2 above)
Comment (jclark) [reply compiler/modules/print.llvm/llvm.bal]: Still doesn't look good to me. Why does addModuleFlag need a DIBuilder?

Comment (heshanpadmasiri) [reply compiler/modules/print.llvm/llvm.bal]: Fixed

**Pull request merged**